### ISSUE TITLE
Feature Request: Derivative-Free Optimization

### ISSUE THREAD

**Issue by bionicles** [tensorflow/tensorflow#23949]
**System information**
- TensorFlow version (you are using): 1.12
- Are you willing to contribute it (Yes/No): Yes
Yes but I'm not sure exactly how to get it done right now.

**Describe the feature and the current behavior/state.**
Often we need to optimize dense nets without derivatives/gradients. It would be excellent to have some built-in derivative-free optimization tools!
https://en.wikipedia.org/wiki/Derivative-free_optimization

**Will this change the current api? How?**
Yes, we could use derivative free optimization without gradients, gradient tape.
There's one example in Tensorflow Probability but the docs aren't super clear on how to implement this:
https://github.com/tensorflow/probability/blob/master/tensorflow_probability/python/optimizer/nelder_mead.py

**Who will benefit with this feature?**
Users who want to do RL can use this if the objective function is difficult / impossible to differentiate, or other situations where gradient calculation doesn't work.

**Any Other info.**
Eager + tf.keras.Model support would be excellent

**Comment by bionicles** [comment]
Here's a review on methods for derivative free optimization:

http://thales.cheme.cmu.edu/dfo/comparison/dfo.pdf

**Comment by bionicles** [comment]
One potential strategy scales well as proven by OpenAI
https://blog.openai.com/evolution-strategies/

**Comment by bionicles** [comment]
Evolving Stable Strategies by David Ha
http://blog.otoro.net/2017/11/12/evolving-stable-strategies/

**Comment by drpngx** [comment]
Sure, that sounds like a good idea. I'm not totally sure if that has enough usage to warrant going into tensorflow itself or if it should be in its standalone repo.

**Comment by alextp** [comment]
This seems like something implementable on top of tensorflow. If you have a prototype it should live somewhere in the ecosystem, but we'd like to wait until something is fairly mature to add it to core tensorflow.

**Comment by bionicles** [comment]
http://www.jmlr.org/papers/volume15/wierstra14a/wierstra14a.pdf

**Comment by iromani** [comment]
I stumbled in a problem of time series forecast to which it would be good to have a custom loss function which depends on iterating over the series sequentially and using state-dependent logic to calculate an accumulation function. I couldn't find a way to do it with TensorFlow symbolic operations, thus my intended loss function is non-differentiable by TensorFlow and hence cannot be used.
